### ISSUE TITLE
[ElasticSearch] store metadata as structured object

### ISSUE THREAD

**Issue by jnioche** [DigitalPebble/storm-crawler#197]
We currently put everything into a single line with tab separated values but could use a map of objects which would allow to query on the metadata values.

**Comment by jnioche** [comment]
Doesn't have to be a nested object but could simply be the use of a common prefix e.g. 'metadata' with [dynamic mapping](https://www.elastic.co/guide/en/elasticsearch/guide/current/custom-dynamic-mapping.html)

**Comment by jnioche** [comment]
First version in 
[metadataAsMap c0a7da3] store metadata as structured object

Need to find how to define properties for any field within the 'metadata' object.

**Comment by jorgelbg** [comment]
With the current implementation the metadata gets indexed as an `object` field if we rely on dynamic mapping right? At least looks that way on a small test I did: 

```json
"_source": {
   "url": "http://addons.firefoxmania.uci.cu",
   "status": "DISCOVERED",
   "metadata": {
      "url.path": [
         "http://firefoxmania.uci.cu"
      ],
      "depth": [
         "1"
      ]
   },
   "nextFetchDate": "2015-10-13T19:33:25.902Z"
}
```

couldn't this cause some sort of error if the values in the `metadata` object are multivalued? as described in [object arrays](https://www.elastic.co/guide/en/elasticsearch/guide/current/complex-core-fields.html#object-arrays) and [nested objects](https://www.elastic.co/guide/en/elasticsearch/guide/current/nested-objects.html)? This could be fixed by forcing metadata to be a `nested` like you suggested initially. 

Using the custom dynamic mapping would mean to "flatten" the metadata object? getting each field with something like `metadata.fieldname` ?

**Comment by jnioche** [comment]
@jorgelbg the metadata values are backed by an array 

```
builder.array(mdKey, values);
```

so this will be fine when there are multiple values.

Having a nested would be an overkill I think.

> Using the custom dynamic mapping would mean to "flatten" the metadata object? getting each field with something like metadata.fieldname ?

I don't think so. Apparently there is a  `path_match` which we could use with e.g. `metadata.*` as a value so that we could make those fields indexable but not analysed nor stored.

This should be a lot better than the current storage we have where we put all the metadata key values on a single line with tab separations.

**Comment by jorgelbg** [comment]
@jnioche Indeed is a lot better than before.

Actually using the default `object` type give some "flat" object style, for instance, in the document I posted, I can issue a query against a `metadata.depth` or `metadata.url.path`, taking a closer look on [inner objects](https://www.elastic.co/guide/en/elasticsearch/guide/current/complex-core-fields.html#_how_inner_objects_are_indexed) shows that basically when an `object` type is used, Elasticsearch flattens internally the structure, otherwise using the `nested` type cause some internal documents to be created.

**Comment by jnioche** [comment]
Implemented in #201
